doschema/jsvalidation.py

```diff
@@ -0,0 +1,316 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of DoSchema.
+# Copyright (C) 2016 CERN.
+#
+# DoSchema is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# DoSchema is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with DoSchema; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+import collections
+import logging
+
+import jsonschema
+import six
+
+from doschema.errors import JSONSchemaCompatibilityException, \
+    JSONSchemaDependenciesException, JSONSchemaEnumException, \
+    JSONSchemaRootNotObjectException, NotSupportedException
+
+
+class JSONSchemaValidation(object):
+    """Class for checking compatibility between schemas."""
+
+    COLLECTION_KEYS = frozenset(['allOf', 'anyOf', 'oneOf'])
+
+    def __init__(self, ignore_index=True, resolver_factory=False,
+                 allow_none_type=False):
+        """Constructor.
+
+        :param ignore_index: If set to True, which is default, it will ignore
+                             array indexes.
+        :param resolver_factory: Resolver used to retrieve referenced schemas.
+                                 If not provided it will use default.
+        :param allow_none_type:
+        """
+        self.ignore_index = ignore_index
+        self.allow_none_type = allow_none_type
+        self.fields_types_dict = {}
+        self.uri = None
+        self.resolver_factory = resolver_factory or jsonschema.RefResolver
+
+    def start(self, schema, uri):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param uri: Name of schema that is currently processed.
+        """
+        self.uri = uri
+        self.resolver = self.resolver_factory(base_uri=uri, referrer=schema)
+
+        if (schema.get('type') != 'object' and 'properties' not in schema and
```

Comment:
There is no need for complicated code in order to check that the root node is of type "object". It can be done by simply checking that `self.fields_types_dict[()] == 'object'` after calling `self.validate_schema(schema, ())`.